File: android2/app/src/main/java/com/xbot/xbot/voice/StreamingTtsPlayer.java
package com.xbot.xbot.voice;

import android.media.AudioAttributes;
import android.media.AudioFormat;
import android.media.AudioTrack;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Streams PCM16 chunks to {@link AudioTrack} for low-latency TTS playback.
 */
public class StreamingTtsPlayer {
    private static final String TAG = "StreamingTts";

    private final AudioCaptureService audio;
    private final ExecutorService playback = Executors.newSingleThreadExecutor();
    private final BlockingQueue<byte[]> queue = new ArrayBlockingQueue<>(256);
    private final AtomicBoolean active = new AtomicBoolean(false);
    private final AtomicBoolean feedingDone = new AtomicBoolean(false);

    @Nullable private AudioTrack track;
    private int sampleRate = 22050;
    private double lastLevel;

    public StreamingTtsPlayer(AudioCaptureService audio) {
        this.audio = audio;
    }

    public boolean isActive() {
        return active.get();
    }

    public void start(int sampleRate) {
        if (active.get()) {
            return;
        }
        this.sampleRate = sampleRate;
        queue.clear();
        feedingDone.set(false);
        audio.setExternalLevel(true);

        int minBuf = AudioTrack.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_OUT_MONO,
                AudioFormat.ENCODING_PCM_16BIT);
        int bufferSize = Math.max(minBuf, sampleRate / 5 * 2);

        track = new AudioTrack.Builder()
                .setAudioAttributes(new AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build())
                .setAudioFormat(new AudioFormat.Builder()
                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                        .setSampleRate(sampleRate)
                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                        .build())
                .setBufferSizeInBytes(bufferSize)
                .setTransferMode(AudioTrack.MODE_STREAM)
                .build();
        track.play();
        active.set(true);
        playback.execute(this::drainLoop);
    }

    public void feedChunk(byte[] pcm16) {
        if (!active.get() || pcm16.length == 0) {
            return;
        }
        queue.offer(pcm16);
    }

    public void markFeedingDone() {
        feedingDone.set(true);
    }

    public boolean waitForDone(long timeoutMs) throws InterruptedException {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (active.get() && System.currentTimeMillis() < deadline) {
            if (feedingDone.get() && queue.isEmpty()) {
                Thread.sleep(50);
                if (queue.isEmpty()) {
                    return true;
                }
            }
            Thread.sleep(20);
        }
        return !active.get();
    }

    public void release() {
        if (!active.compareAndSet(true, false)) {
            resetLevel();
            return;
        }
        queue.clear();
        AudioTrack local = track;
        track = null;
        if (local != null) {
            try {
                local.stop();
            } catch (IllegalStateException ignored) {
            }
            local.release();
        }
        resetLevel();
    }

    private void drainLoop() {
        while (active.get()) {
            try {
                byte[] chunk = queue.poll(100, TimeUnit.MILLISECONDS);
                if (chunk == null) {
                    if (feedingDone.get() && queue.isEmpty()) {
                        break;
                    }
                    continue;
                }
                AudioTrack local = track;
                if (local == null) {
                    break;
                }
                updateLevel(chunk);
                local.write(chunk, 0, chunk.length);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        Log.d(TAG, "drain loop finished");
    }

    private void updateLevel(byte[] chunk) {
        float raw = AudioCaptureService.rmsLevel(chunk);
        lastLevel = lastLevel * 0.6 + raw * 0.4;
        audio.postLevel((float) lastLevel);
    }

    private void resetLevel() {
        audio.setExternalLevel(false);
        audio.postLevel(0f);
        lastLevel = 0;
    }

    public void dispose() {
        release();
        playback.shutdownNow();
    }
}

File: xbot-native/app/src/main/java/com/xbot/android/voice/StreamingTtsPlayer.kt
package com.xbot.android.voice

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.util.Log
import java.util.concurrent.LinkedBlockingDeque

/**
 * 流式 TTS 播放器：用 AudioTrack 喂 PCM16 chunk，首个 chunk 到达即开口。
 * 对应 Flutter streaming_tts_player.dart（flutter_pcm_sound 的原生等价）。
 *
 * - PCM16 / mono / streaming mode，采样率由 meta 给（默认 22050）
 * - 首 chunk 即播放（首音延迟 ≈ LLM done + DashScope 首包）
 * - 喂 chunk 时同步写 [onLevel]（嘴部同步，RMS/6000）
 * - 等队列耗尽 + AudioTrack 自然播完（drain）再 release，避免截断
 *
 * @param onLevel 实时音量 0..1（驱动虚拟形象嘴部）；TTS 期间接管 AudioCapture.level
 */
class StreamingTtsPlayer(
    private val onLevel: (Float) -> Unit,
) {
    companion object {
        private const val TAG = "StreamingTtsPlayer"
        private const val RMS_DENOM = 6000f
    }

    private var track: AudioTrack? = null
    private var sampleRate = 22050
    private val queue = LinkedBlockingDeque<ShortArray>()
    @Volatile private var feedingDone = false
    @Volatile private var active = false
    private var feedThread: Thread? = null
    @Volatile var totalSamplesFed = 0
        private set
    @Volatile var drained = false
        private set

    /** 开始：初始化 AudioTrack（采样率由 meta 决定，首次 feedChunk 前调 start）。 */
    fun start(sr: Int = 22050) {
        if (active) return
        sampleRate = sr
        val minBuf = AudioTrack.getMinBufferSize(
            sampleRate, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT
        )
        val bufSize = maxOf(minBuf, sampleRate / 5) // 至少 ~200ms 缓冲
        track = AudioTrack(
            AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_MEDIA)
                .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                .build(),
            AudioFormat.Builder()
                .setSampleRate(sampleRate)
                .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                .build(),
            bufSize,
            AudioTrack.MODE_STREAM,
            AudioManager.AUDIO_SESSION_ID_GENERATE,
        )
        track!!.play()
        active = true
        feedingDone = false
        drained = false
        totalSamplesFed = 0
        feedThread = Thread({ runFeedLoop() }, "xbot-tts-feed").apply { isDaemon = true; start() }
    }

    /** 喂一个 PCM16 chunk（little-endian bytes → short[]）。立即写入队列触发播放。 */
    fun feedChunk(pcm16: ByteArray) {
        if (!active) return
        val samples = bytesToShorts(pcm16)
        queue.add(samples)
    }

    /** 标记喂入完成（所有 chunk 已喂完）；等播放耗尽后触发 drain。 */
    fun markFeedingDone() {
        feedingDone = true
    }

    private fun runFeedLoop() {
        try {
            while (active) {
                val chunk = queue.poll() ?: run {
                    if (feedingDone) break
                    Thread.sleep(5); continue
                }
                totalSamplesFed += chunk.size
                // 嘴部同步：写音量。
                onLevel(rmsLevel(chunk))
                track?.write(chunk, 0, chunk.size)
            }
            // 等 AudioTrack 把硬件缓冲播完（drain）。
            track?.stop()
            track?.flush()
            drained = true
            onLevel(0f)
        } catch (e: Exception) {
            Log.e(TAG, "feed 异常: ${e.message}")
        }
    }

    /** 释放：停止播放 + 清队列 + 释放 AudioTrack。 */
    fun release() {
        active = false
        queue.clear()
        try { feedThread?.join(300) } catch (_: Exception) {}
        feedThread = null
        try { track?.stop() } catch (_: Exception) {}
        track?.release()
        track = null
        onLevel(0f)
    }

    private fun bytesToShorts(bytes: ByteArray): ShortArray {
        val n = bytes.size / 2
        val out = ShortArray(n)
        for (i in 0 until n) {
            val lo = bytes[i * 2].toInt() and 0xFF
            val hi = bytes[i * 2 + 1].toInt()
            out[i] = ((hi shl 8) or lo).toShort()
        }
        return out
    }

    private fun rmsLevel(samples: ShortArray): Float {
        var sumSq = 0.0
        for (s in samples) { val v = s.toDouble(); sumSq += v * v }
        val rms = kotlin.math.sqrt(sumSq / samples.size)
        return (rms / RMS_DENOM).toFloat().coerceIn(0f, 1f)
    }
}
